File: Dynamic_Programming/3543. Maximum Weighted K-Edge Path.py
# 遞歸 如果要寫成遞推可以用set才不會超時 TC:O((n+m)kt) SC:O(nkt) //  Topological Sort + 狀態: TC:O((n+m)kt) SC:O(nkt) // 遞推DP 且更改狀態 只需以edges 遍歷 TC:O(mkt) SC:O(nkt)
class Solution:
    def maxWeight(self, n: int, edges: List[List[int]], k: int, t: int) -> int:
        # 遞推DP 且更改狀態 只需以edges 遍歷 如若要用bitset優化需要用c++
        f = [[set() for _ in range(n)] for _ in range(k + 1)]
        for i in range(n):
            f[0][i].add(0) # 剛開始都可以走0步然後有sum = 0
        for step in range(k): # 走了幾步
            for u, v, w in edges: # 當前節點
                for uw in f[step][u]:
                    if uw + w < t:
                        f[step + 1][v].add(uw + w)
        return max((max(st) for st in f[k] if st), default=-1)
    
        '''
        Topological Sort + 狀態
        graph = [[] for _ in range(n)]
        deg = [0] * n
        for e in edges:
            graph[e[0]].append((e[1], e[2]))
            deg[e[1]] += 1
        f = [[set() for _ in range(k + 1)] for _ in range(n)]
        q = deque(i for i, d in enumerate(deg) if d == 0)
        ans = -1
        while q:
            x = q.popleft()
            f[x][0].add(0) # 這點開始的狀態
            if f[x][k]:
                ans = max(ans, max(f[x][k]))
            for y, w in graph[x]: # 之後要走到的位置與權重
                for i in range(k): # 當前可能走過的edges數
                    for xw in f[x][i]:
                        if xw + w < t:
                            f[y][i + 1].add(xw + w)
                deg[y] -= 1
                if deg[y] == 0:
                    q.append(y)
        return ans
        '''

        '''
        遞歸
        graph = [[] for _ in range(n)]
        for e in edges:
            graph[e[0]].append((e[1], e[2]))
        ans = -1
        @cache
        def dfs(i:int, cnt:int, tot:int) -> None:
            if cnt == k:
                nonlocal ans
                if cnt == k:
                    ans = max(ans, tot)
                return
            for nxt, w in graph[i]:
                if w + tot < t:
                    dfs(nxt, cnt + 1, tot + w)
        for i in range(n):
            dfs(i, 0, 0)
        return ans
        '''

File: Dynamic_Programming/3543. Maximum Weighted K-Edge Path.cpp
// 遞歸 如果要寫成遞推可以用set才不會超時 TC:O((n+m)kt) SC:O(nkt) Bitset: TC:O(mkt/w) m 為 edges數量 SC:O(nkt/w)
// 似乎還可以用拓墣排序做
static const int MAXSTEP = 601;
class Solution {
public:
    int maxWeight(int n, vector<vector<int>>& edges, int k, int t) {
        // Bitset作法
        vector f(k + 1, vector<bitset<600>>(n));
        for (int i = 0; i < n; i++) {
            f[0][i].set(0);
        }
        for (int i = 0; i < k; i++) {
            for (auto& e : edges) {
                int x = e[0], y = e[1], wt = e[2];
                f[i + 1][y] |= f[i][x] << wt;
            }
        }

        int ans = -1;
        for (auto& bit : f[k]) {
            for (int s = t - 1; s >= 0; s--) {
                if (bit.test(s)) {
                    ans = max(ans, s);
                    break;
                }
            }
        }
        return ans;
        
        /*
        vector<vector<pair<int,int>>> next(n);
        for (auto &e : edges) {
            next[e[0]].emplace_back(e[1], e[2]);
        }
        
        vector<unordered_map<int, int>> memo(n * (k+1));
        
        auto dfs = [&](this auto&&dfs, int node, int cnt, int sum) -> int {
            if (sum >= t) return -1;
            if (cnt == k) return sum;
            int idx = node * (k+1) + cnt;
            
            // auto it = memo[idx].find(sum);
            if (memo[idx].find(sum) != memo[idx].end()) {
                return memo[idx][sum];
            }
        
            int ans = -1;
            for (auto [nxt, w] : next[node]) {
                ans = max(ans, dfs(nxt, cnt + 1, sum + w));
            }
            
            memo[idx][sum] = ans;
            return ans;
        };
        int answer = -1;
        for (int i = 0; i < n; i++) {
            answer = max(answer, dfs(i, 0, 0));
        }
        
        return answer;
        */
    }
};
